### PR TITLE
feat(bigtable): add async sample rows

### PR DESCRIPTION
**Pull request by dbolduc**
Fixes: #1490

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/6561)

### CODE REVIEW

File: google/cloud/bigtable/data_client.cc [hunk 1]
```diff
@@ -139,6 +139,14 @@ class DefaultDataClient : public DataClient {
       ::grpc::CompletionQueue* cq, void* tag) override {
     return impl_.Stub()->AsyncSampleRowKeys(context, request, cq, tag);
   }
+  std::unique_ptr<::grpc::ClientAsyncReaderInterface<
```
Comment (coryan) [inline google/cloud/bigtable/data_client.cc:142]: We talked about this over IM too, but consider returning something like `std::unique_ptr<google::cloud::NewClass<::google::bigtable::v2::SampleRowKeysResponse>>` here.

File: google/cloud/bigtable/data_client.h [hunk 2]
```diff
@@ -166,6 +166,12 @@ class DataClient {
       ::grpc::ClientContext* context,
       const ::google::bigtable::v2::SampleRowKeysRequest& request,
       ::grpc::CompletionQueue* cq, void* tag) = 0;
+  virtual std::unique_ptr<::grpc::ClientAsyncReaderInterface<
+      ::google::bigtable::v2::SampleRowKeysResponse>>
+  PrepareAsyncSampleRowKeys(
+      ::grpc::ClientContext* context,
+      const ::google::bigtable::v2::SampleRowKeysRequest& request,
+      ::grpc::CompletionQueue* cq) = 0;
```
Comment (coryan) [inline google/cloud/bigtable/data_client.h:174]: As you know, this is a breaking change.

File: google/cloud/bigtable/internal/async_row_sampler_test.cc [hunk 3]
```diff
@@ -0,0 +1,449 @@
+// Licensed under the Apache License, Version 2.0 (the "License");
```
Comment (coryan) [inline google/cloud/bigtable/internal/async_row_sampler_test.cc:1]: nit: your boilerplate is missing information.

File: google/cloud/bigtable/internal/async_row_sampler_test.cc [hunk 4]
```diff
@@ -0,0 +1,449 @@
...
+  // Finish()
+  cq_impl_->SimulateCompletion(true);
+
+  ASSERT_EQ(0U, cq_impl_->size());
+
+  auto status = samples_future.get();
+  ASSERT_STATUS_OK(status);
+
+  auto samples = status.value();
+  ASSERT_EQ(1U, samples.size());
```
Comment (coryan) [inline google/cloud/bigtable/internal/async_row_sampler_test.cc:171]: Remember the comments about testing with `ElementsAre` here.

File: google/cloud/bigtable/internal/async_row_sampler_test.cc [hunk 5]
```diff
@@ -0,0 +1,449 @@
...
+  cq_impl_->SimulateCompletion(true);
+
+  auto status = samples_future.get();
+  ASSERT_THAT(status,
+              StatusIs(StatusCode::kUnavailable, HasSubstr("try again")));
+
+  ASSERT_EQ(0U, cq_impl_->size());
+}
+
+class MockBackoffPolicy : public RPCBackoffPolicy {
```
Comment (coryan) [inline google/cloud/bigtable/internal/async_row_sampler_test.cc:230]: Consider refactoring this, it now appears in several tests.... Also move to the top.  If that is too much work for this PR, create an issue to clean this up.

File: google/cloud/bigtable/internal/async_row_sampler_test.cc [hunk 6]
```diff
@@ -0,0 +1,449 @@
...
+  cq_impl_->SimulateCompletion(false);
+  // Finish()
+  cq_impl_->SimulateCompletion(true);
+
+  auto status = samples_future.get();
+  ASSERT_STATUS_OK(status);
+
+  auto samples = status.value();
+  ASSERT_EQ(1U, samples.size());
+  EXPECT_EQ("test1", samples[0].row_key);
```
Comment (coryan) [inline google/cloud/bigtable/internal/async_row_sampler_test.cc:397]: Remember the comments about testing using `ElementsAre`.

**Review by coryan — CHANGES_REQUESTED** [review]
A few changes requested below, mostly nits and improvements to make the tests easier to troubleshoot.

The code looks good, we need to figure out the backwards compatibility stuff.

**Review by dbolduc — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/6561)* status: 0 of 21 files reviewed, 8 unresolved discussions (waiting on @coryan)

___
*[google/cloud/bigtable/data_client.h, line 174 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/6561#-M_aJzEFs81EUgSO6yl8-r1-174:-M_gHyvnBJ-PpeDR6QzU:bfm7bsp) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/6015b0cdf5ee6c26765b38b784aed54bf939359e/google/cloud/bigtable/data_client.h#L174)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

As you know, this is a breaking change.
</blockquote></details>

Discussed over chat: We resolved to make this a non-pure virtual function that returns `nullptr`.

We think this is fine because `PrepareAsyncSampleRowKeys` is a newly added method, so none of our users currently depend on it. If users wish to start using the method in one of their own `DataClient`'s, it will be their responsibility to override this method.

Our next release will include some sort of warning about this change in `CHANGELOG.md`.

Additionally, we provide an implementation for this method in our classes derived from `DataClient` (e.g. `DefaultDataClient`)
___
*[google/cloud/bigtable/data_client.cc, line 142 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/6561#-M_aJzEFs81EUgSO6yl7-r1-142:-M_gU5tN41Rtbvz3MVU3:bdirjqq) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/6015b0cdf5ee6c26765b38b784aed54bf939359e/google/cloud/bigtable/data_client.cc#L142)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

We talked about this over IM too, but consider returning something like `std::unique_ptr<google::cloud::NewClass<::google::bigtable::v2::SampleRowKeysResponse>>` here.
</blockquote></details>

See comment in data_client.h
___
*[google/cloud/bigtable/table.h, line 706 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/6561#-M_aJzEFs81EUgSO6ylE-r1-706:-M_bnOeD7QzUNoKdktef:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/6015b0cdf5ee6c26765b38b784aed54bf939359e/google/cloud/bigtable/table.h#L706)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

s/Asynchronous sample/Asynchronously obtains a sample/
</blockquote></details>

Done.
___
*[google/cloud/bigtable/table.h, line 716 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/6561#-M_aJzEFs81EUgSO6ylE-r1-716:-M_bnQheCGLe_rOuaho2:bhdpljr) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/6015b0cdf5ee6c26765b38b784aed54bf939359e/google/cloud/bigtable/table.h#L716)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

That seems wrong, this is a read-only operation, we should always treat it as idempotent.
</blockquote></details>

Good catch. Description of `SampleRows()` has been updated as well.
___
*[google/cloud/bigtable/internal/async_row_sampler_test.cc, line 171 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/6561#-M_aJzEFs81EUgSO6ylM-r1-171:-M_gk7Hg8ZfJTVDseLKU:b-ow41nf) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/6015b0cdf5ee6c26765b38b784aed54bf939359e/google/cloud/bigtable/internal/async_row_sampler_test.cc#L171)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

Remember the comments about testing with `ElementsAre` here.
</blockquote></details>

Done. Thank you for explaining (twice) why we want to do this.
___
*[google/cloud/bigtable/internal/async_row_sampler_test.cc, line 230 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/6561#-M_aJzEFs81EUgSO6ylM-r1-230:-M_d808rDFCipHgMBPMZ:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/6015b0cdf5ee6c26765b38b784aed54bf939359e/google/cloud/bigtable/internal/async_row_sampler_test.cc#L230)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

Consider refactoring this, it now appears in several tests.... Also move to the top.  If that is too much work for this PR, create an issue to clean this up.
</blockquote></details>

Done.
___
*[google/cloud/bigtable/internal/async_row_sampler_test.cc, line 397 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/6561#-M_aJzEFs81EUgSO6ylM-r1-397:-M_gkEd01VWtdut2LlRT:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/6015b0cdf5ee6c26765b38b784aed54bf939359e/google/cloud/bigtable/internal/async_row_sampler_test.cc#L397)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

Remember the comments about testing using `ElementsAre`.
</blockquote></details>

Done.
___
*[google/cloud/bigtable/internal/logging_data_client.cc, line 149 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/6561#-M_aJzEFs81EUgSO6ylB-r1-149:-M_d86TKAsEwn-qJ6Tf8:bbvia0c) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/6015b0cdf5ee6c26765b38b784aed54bf939359e/google/cloud/bigtable/internal/logging_data_client.cc#L149)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

No action needed.  This is why we no longer like returning a raw `std::unique_ptr<grpc::ClientAsyncReaderInterface<...>>`.  We typically wrap those in our own class so we (and our customers) can mock them and decorate them.  Unfortunately, we only have wrappers for blocking streams, and for asynchronous bidirectional streams.  This is an area where you should start thinking of improvements to Bigtable.
</blockquote></details>

Ack. Thanks for pointing this out.

File: google/cloud/bigtable/internal/logging_data_client.cc [hunk 7]
```diff
@@ -140,6 +140,15 @@ LoggingDataClient::AsyncSampleRowKeys(
   return child_->AsyncSampleRowKeys(context, request, cq, tag);
 }
 
+std::unique_ptr<::grpc::ClientAsyncReaderInterface<
+    ::google::bigtable::v2::SampleRowKeysResponse>>
+LoggingDataClient::PrepareAsyncSampleRowKeys(
+    ::grpc::ClientContext* context,
+    const ::google::bigtable::v2::SampleRowKeysRequest& request,
```
Comment (devjgm) [inline google/cloud/bigtable/internal/logging_data_client.cc:147]: Oops. An accidental west-const here. As you know, we use east-const (`const int` == "west const"; `int const` == "east const").

(I wish there was a tool to fix this. There's discussion of teaching clang-format how to fix this, but that has been stalled for a couple years.)

**Review by devjgm — COMMENTED** [review]
Ignore my "east const" comment. I see there a bunch of those that we could clean up later in a separate PR.

**Review by coryan — COMMENTED** [review]
A few nits, and few include paths and style that need fixing.  Otherwise this is looking good.

> Reviewed 11 of 18 files at r1, 10 of 10 files at r2.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/6561#-:-M_gv50U-TGCxdAhN6UC:b-uym8jr)* status: all files reviewed, 7 unresolved discussions (waiting on @dbolduc and @devjgm)

___
*[google/cloud/bigtable/data_client.h, line 174 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/6561#-M_aJzEFs81EUgSO6yl8-r1-174:-M_gsZRX3pF0bTlchLFZ:bb6ywmx) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/6015b0cdf5ee6c26765b38b784aed54bf939359e/google/cloud/bigtable/data_client.h#L174)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

Discussed over chat: We resolved to make this a non-pure virtual function that returns `nullptr`.

We think this is fine because `PrepareAsyncSampleRowKeys` is a newly added method, so none of our users currently depend on it. If users wish to start using the method in one of their own `DataClient`'s, it will be their responsibility to override this method.

Our next release will include some sort of warning about this change in `CHANGELOG.md`.

Additionally, we provide an implementation for this method in our classes derived from `DataClient` (e.g. `DefaultDataClient`)
</blockquote></details>

LGTM
___
*[google/cloud/bigtable/data_client.cc, line 142 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/6561#-M_aJzEFs81EUgSO6yl7-r1-142:-M_gshQp2CN0_Ek3XVa-:b-81nlyc) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/6015b0cdf5ee6c26765b38b784aed54bf939359e/google/cloud/bigtable/data_client.cc#L142)):*
<details><summary><i>Previously, dbolduc (Darren Bolduc) wrote…</i></summary><blockquote>

See comment in data_client.h
</blockquote></details>

SGTM.
___
*[google/cloud/bigtable/data_client.cc, line 152 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/6561#-M_gseC608y67L09mmTz:-M_gseC7EMTWH5adLiLI:b-vn8j26) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/d7f406d5465c56c10fd0d88c6bd4983edfe10d8c/google/cloud/bigtable/data_client.cc#L152)):*
> ```cpp
>     return impl_.Stub()->AsyncSampleRowKeys(context, request, cq, tag);
>   }
>   std::unique_ptr<::grpc::ClientAsyncReaderInterface<
> ```

nit: blank line before starting the function.
___
*[google/cloud/bigtable/table.cc, line 22 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/6561#-M_gsuaB7gPCJrgmBD0e:-M_gsuaB7gPCJrgmBD0f:b-w0ohag) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/d7f406d5465c56c10fd0d88c6bd4983edfe10d8c/google/cloud/bigtable/table.cc#L22)):*
> ```cpp
> #include "google/cloud/grpc_error_delegate.h"
> #include "google/cloud/internal/async_retry_unary_rpc.h"
> #include "data_client.h"
> ```

include paths should be fully qualified, I am not sure why you need them also.
___
*[google/cloud/bigtable/internal/async_row_sampler_test.cc, line 1 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/6561#-M_gtJgFEilBJLCAyf9w:-M_gtJgFEilBJLCAyf9x:b-w9u095) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/d7f406d5465c56c10fd0d88c6bd4983edfe10d8c/google/cloud/bigtable/internal/async_row_sampler_test.cc#L1)):*
> ```cpp
> // Licensed under the Apache License, Version 2.0 (the "License");
> ```

Still missing a portion of this boiler plate.
___
*[google/cloud/bigtable/internal/async_row_sampler_test.cc, line 19 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/6561#-M_gtN2x8fw-agh10x_g:-M_gtN2x8fw-agh10x_h:b-vbst3t) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/d7f406d5465c56c10fd0d88c6bd4983edfe10d8c/google/cloud/bigtable/internal/async_row_sampler_test.cc#L19)):*
> ```cpp
> #include "google/cloud/testing_util/fake_completion_queue_impl.h"
> #include "google/cloud/testing_util/status_matchers.h"
> #include "google/bigtable/v2/bigtable.pb.h"
> ```

includes for proto-generated files should be with angle brackets `#include <google/bigtable/v2/bigtable.pb.h>` that suppresses a lot of spurious warnings with some compilers / tools.
___
*[google/cloud/bigtable/internal/logging_data_client.cc, line 147 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/6561#-M_aJzEFs81EUgSO6ylB-r2-147:-M_gudz_DkEgMA6gadpQ:bqhu35u) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/d7f406d5465c56c10fd0d88c6bd4983edfe10d8c/google/cloud/bigtable/internal/logging_data_client.cc#L147)):*
<details><summary><i>Previously, devjgm (Greg Miller) wrote…</i></summary><blockquote>

Oops. An accidental west-const here. As you know, we use east-const (`const int` == "west const"; `int const` == "east const").

(I wish there was a tool to fix this. There's discussion of teaching clang-format how to fix this, but that has been stalled for a couple years.)
</blockquote></details>

Please file a bug to fix the "east" vs. "west" const-ness of these files.
___
*[google/cloud/bigtable/testing/inprocess_data_client.cc, line 147 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/6561#-M_guruP0gUOGtj1vxTp:-M_guruP0gUOGtj1vxTq:b-qpw9go) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/d7f406d5465c56c10fd0d88c6bd4983edfe10d8c/google/cloud/bigtable/testing/inprocess_data_client.cc#L147)):*
> ```cpp
>   return Stub()->PrepareAsyncSampleRowKeys(context, request, cq);
> }
> }  // namespace testing
> ```

nit: preserve the blank line before closing the namespaces.
___
*[google/cloud/bigtable/testing/mock_backoff_policy.h, line 20 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/6561#-M_guJ2I8-vk2y6THCqs:-M_guJ2J8O_AQ4PjYuOz:bktdtgy) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/d7f406d5465c56c10fd0d88c6bd4983edfe10d8c/google/cloud/bigtable/testing/mock_backoff_policy.h#L20)):*
> ```objc
> #include "google/cloud/bigtable/rpc_backoff_policy.h"
> #include "google/cloud/status.h"
> #include "grpcpp/client_context.h"
> ```

In general, includes for "other projects" are with angle brackets. Abseil is the sad exception for reasons too complicated to put in this margin.

**Review by dbolduc — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/6561)* status: 12 of 21 files reviewed, 5 unresolved discussions (waiting on @coryan and @devjgm)

___
*[google/cloud/bigtable/table.cc, line 22 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/6561#-M_gsuaB7gPCJrgmBD0e:-M_h7SFN5AkZsDhi7hsy:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/d7f406d5465c56c10fd0d88c6bd4983edfe10d8c/google/cloud/bigtable/table.cc#L22)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

include paths should be fully qualified, I am not sure why you need them also.
</blockquote></details>

Done.
___
*[google/cloud/bigtable/internal/async_row_sampler_test.cc, line 1 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/6561#-M_gtJgFEilBJLCAyf9w:-M_h7fL29XlBHcdN0QgX:b1lsdgt) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/d7f406d5465c56c10fd0d88c6bd4983edfe10d8c/google/cloud/bigtable/internal/async_row_sampler_test.cc#L1)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

Still missing a portion of this boiler plate.
</blockquote></details>

Embarrassing. Done.
___
*[google/cloud/bigtable/internal/async_row_sampler_test.cc, line 19 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/6561#-M_gtN2x8fw-agh10x_g:-M_h8MXY0o0HqR98ceiW:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/d7f406d5465c56c10fd0d88c6bd4983edfe10d8c/google/cloud/bigtable/internal/async_row_sampler_test.cc#L19)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

includes for proto-generated files should be with angle brackets `#include <google/bigtable/v2/bigtable.pb.h>` that suppresses a lot of spurious warnings with some compilers / tools.
</blockquote></details>

Done.
___
*[google/cloud/bigtable/internal/logging_data_client.cc, line 147 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/6561#-M_aJzEFs81EUgSO6ylB-r2-147:-M_h8OQg31isefdLScBK:bgjiboo) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/d7f406d5465c56c10fd0d88c6bd4983edfe10d8c/google/cloud/bigtable/internal/logging_data_client.cc#L147)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

Please file a bug to fix the "east" vs. "west" const-ness of these files.
</blockquote></details>

@devjgm will fix the pre-existing west const's. I will not introduce new west const's.
___
*[google/cloud/bigtable/testing/mock_backoff_policy.h, line 20 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/6561#-M_guJ2I8-vk2y6THCqs:-M_hC-wM2BfMPO7xG-4e:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/d7f406d5465c56c10fd0d88c6bd4983edfe10d8c/google/cloud/bigtable/testing/mock_backoff_policy.h#L20)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

In general, includes for "other projects" are with angle brackets. Abseil is the sad exception for reasons too complicated to put in this margin.
</blockquote></details>

Done.

**Comment by dbolduc** [comment]
looks like: #6266

**Review by coryan — APPROVED** [review]
:lgtm:

> Reviewed 9 of 9 files at r3.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/6561#-:-M_kHcCzECldiqdN1oiA:bnfp4nl)* status: all files reviewed, 1 unresolved discussion (waiting on @devjgm)

**Comment by coryan** [comment]
I restarted the build that flaked out, feel free to do that yourself in the future (I can show you how over a GVC or something, hard to explain in text).
You should be able to merge even thought Reviewable is waiting for @devjgm to click some button to close out his comments, he was Okay with the changes anyway, the tool is not smart enough to tell.

**Pull request merged**